x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java

```diff
@@ -717,7 +720,10 @@ public Literal visitParamLiteral(ParamLiteralContext ctx) {
         }
         // otherwise we need to make sure that xcontent-serialized value is converted to the correct type
         try {
-            return new Literal(source, converterFor(sourceType, dataType).convert(param.value), dataType);
+            if (SqlDataTypeConverter.canConvert(sourceType, dataType) == false) {
+                throw new ParsingException(source, "Can not cast value [{}] as [{}]", param.value, dataType);
```

Comment:
I think we should rephrase here, the `cast` ca be misleading since there is not actual `CAST(<xxx>  AS <yyy>)`.
Maybe a message along the lines of: `Cannot bind value [{}] of type [{}] to parameter of type [{}].

Please also add a test case for this.

Reply:
Thanks. I've added a test and updated the message.
I've kept the `cast` though to stay consistent with the (pre-existing) message that's return when a converter actually fails to convert a value (vs. no converter exists, as in this case). 
Besides, the confusion should be minimal, type "casting" is a coined term, `CAST` function is a very specific functionality.

Reply:
Since the error message refers to binding a value to a param `?`, 
imho, `bind` sounds better than `cast`.